### PR TITLE
feat(join): refactor and support outer join for hash join

### PR DESCRIPTION
**Pull request by wangrunji0408**
This PR refactors `NestedLoopJoin` and `HashJoin` executors, and adds outer join support to hash join.

### CODE REVIEW

File: src/executor/nested_loop_join.rs [hunk 1]
```diff
@@ -21,170 +21,94 @@ pub struct NestedLoopJoinExecutor {
...
+            ArrayImpl::Bool(a) => a,
+            _ => panic!("unsupported value from join condition"),
         };
+        yield cross_chunk.filter(filter.iter().map(|b| matches!(b, Some(true))));
 
-        for left_chunk_idx in 0..left_chunks.len() {
-            for left_row_idx in 0..left_chunks[left_chunk_idx].cardinality() {
+        // append rows for left outer join
+        if matches!(
+            self.join_op,
```
Comment (TennyZhuang) [inline src/executor/nested_loop_join.rs:62]: I prefer treat join_op as const generics, but we can refine it later.

File: src/executor/nested_loop_join.rs [hunk 2]
```diff
@@ -21,170 +21,94 @@ pub struct NestedLoopJoinExecutor {
...
+            BoundJoinOperator::RightOuter | BoundJoinOperator::FullOuter
+        ) {
+            let mut builders = create_builders();
+            // scan the filter to find whether a right row matches any left rows
+            let mut matched = bitvec![0; right_rows().count()];
+            let mut i = 0;
+            for _ in left_rows() {
+                for (j, _) in right_rows().enumerate() {
```
Comment (TennyZhuang) [inline src/executor/nested_loop_join.rs:96]: Why not 0..right_rows().len()?

File: src/executor/nested_loop_join.rs [hunk 3]
```diff
@@ -21,170 +21,94 @@ pub struct NestedLoopJoinExecutor {
...
+                    if matches!(filter.get(i), Some(true)) {
+                        matched.set(j, true);
                     }
+                    i += 1;
```
Comment (TennyZhuang) [inline src/executor/nested_loop_join.rs:100]: Why maintain I manually?

File: src/executor/nested_loop_join.rs (hunk 2 above)
Comment (wangrunji0408) [reply src/executor/nested_loop_join.rs:96]: Good idea...

File: src/executor/nested_loop_join.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/executor/nested_loop_join.rs:100]: I need something like `.enumerate()` in the nested for-loop. This way is the simplest I can think (

Comment (TennyZhuang) [reply src/executor/nested_loop_join.rs:100]: got it

Comment (TennyZhuang) [reply src/executor/nested_loop_join.rs:100]: How about https://docs.rs/itertools/0.4.8/itertools/trait.Itertools.html#method.cartesian_product?

Comment (TennyZhuang) [reply src/executor/nested_loop_join.rs:100]: catesian_product().enumerate(), then we may got i?

Not necessary, any one is ok

File: src/executor/nested_loop_join.rs (hunk 1 above)
Comment (wangrunji0408) [reply src/executor/nested_loop_join.rs:62]: Sounds good. But if we do that, we have to construct 4 types when building the executor...

Comment (TennyZhuang) [reply src/executor/nested_loop_join.rs:62]: Yes, @yuhao-su has done it in another repo.

Comment (TennyZhuang) [reply src/executor/nested_loop_join.rs:62]: We can do it after adt_const_generics completed.

File: src/executor/nested_loop_join.rs (hunk 3 above)
Comment (wangrunji0408) [reply src/executor/nested_loop_join.rs:100]: I tried:
```rust
for (i, (_, (j, _))) in left_rows()
    .cartesian_product(right_rows().enumerate())
    .enumerate()
{
    if matches!(filter.get(i), Some(true)) {
        matched.set(j, true);
    }
}
```

However, `.cartesian_product(other)` requires `other` to be cloneable, unfortunately that is impossible in our case.

Give up 🤪

Comment (wangrunji0408) [reply src/executor/nested_loop_join.rs:100]: I'm stupid. It works now:
```rust
for (i, (_, j)) in left_rows()
    .cartesian_product(0..right_rows().count())
    .enumerate()
{
    if matches!(filter.get(i), Some(true)) {
        matched.set(j, true);
    }
}
```

Comment (wangrunji0408) [reply src/executor/nested_loop_join.rs]: I found another way which can avoid bitmap: a46215c

File: src/executor/hash_join.rs [hunk 4]
```diff
@@ -13,89 +15,91 @@ pub struct HashJoinExecutor {
...
-            chunk_builders.push(ArrayBuilderImpl::new(ty));
+
+        // probe
+        let mut builders = (self.left_types.iter())
+            .chain(self.right_types.iter())
+            .map(|ty| ArrayBuilderImpl::with_capacity(PROCESSING_WINDOW_SIZE, ty))
+            .collect_vec();
+        for right_row in right_rows() {
+            let hash_value = right_row.get(self.right_column_index);
+            for left_row in hash_map.get(&hash_value).unwrap_or(&vec![]) {
```
Comment (skyzh) [inline src/executor/hash_join.rs:54]: I would prefer `unwrap_or_else` or create the empty vector once at the beginning of this function.

File: src/executor/hash_join.rs [hunk 5]
```diff
@@ -13,89 +15,91 @@ pub struct HashJoinExecutor {
...
+                let values = left_row.values().chain(right_row.values());
+                for (builder, v) in builders.iter_mut().zip(values) {
```
Comment (skyzh) [inline src/executor/hash_join.rs:56]: consider using `Itertools::zip_eq`.

**Review by skyzh — APPROVED** [review]
Rest LGTM, good work!

File: src/executor/hash_join.rs (hunk 4 above)
Comment (wangrunji0408) [reply src/executor/hash_join.rs:54]: `vec![]` (`Vec::new()`) is a const fn. So I guess the compiler should be able to optimize here?

Comment (skyzh) [reply src/executor/hash_join.rs:54]: Get it! Then I think there's no need to optimize :)

**Review by TennyZhuang — APPROVED** [review]
LGTM

**Pull request merged**